### PR TITLE
Automated comments for Java hamming exercise

### PR DESCRIPTION
**Pull request by jmrunkle**
Creates automated comments to go with the latest changes for the Java analyzer. See https://github.com/exercism/java-analyzer/pull/18

### CODE REVIEW

File: automated-comments/java/general/constructor_too_long.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+At least one constructor in this solution is getting long.
```
Comment (tehsphinx) [inline automated-comments/java/general/constructor_too_long.md]: Is it possible to identify the constructor in question and name it? Then this can be written less vague.

File: automated-comments/java/general/method_too_long.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+At least one method in this solution is getting long.
```
Comment (tehsphinx) [inline automated-comments/java/general/method_too_long.md]: Same here: maybe add the name of the method per parameter.

File: automated-comments/java/hamming/must_calculate_hamming_distance.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+There are no loops or lambdas in this solution.
+We recommend either looping or streaming through the indices to calculate the Hamming distance.
+Other solutions tend to be difficult to reason about and difficult read.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/must_calculate_hamming_distance.md]:
```suggestion
Other solutions tend to be difficult to reason about and difficult to read.
```

File: automated-comments/java/hamming/must_throw_in_constructor.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+The tests indicate that the constructor ought to be throwing IllegalArgumentExceptions with specific error messages.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/must_throw_in_constructor.md]:
```suggestion
The tests indicate that the constructor ought to be throwing `IllegalArgumentExceptions` with specific error messages.
```

File: automated-comments/java/hamming/must_throw_in_constructor.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+Make sure the constructor throws if the inputs have different lengths.
+
+NOTE: An IllegalArgumentException is not a checked exception so it does not need a `throws` statement on the constructor (or any other methods).
```
Comment (tehsphinx) [inline automated-comments/java/hamming/must_throw_in_constructor.md]:
```suggestion
NOTE: An `IllegalArgumentException` is not a checked exception so it does not need a `throws` statement on the constructor (or any other methods).
```

File: automated-comments/java/hamming/must_use_conditional_logic_in_constructor.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Since the constructor is required to use a different message depending on the input, there should be conditional logic inside the constructor's execution path.
+Either the constructor, or methods called by the constructor, should use if statements.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/must_use_conditional_logic_in_constructor.md]:
```suggestion
Either the constructor, or methods called by the constructor, should use `if` statements.
```

File: automated-comments/java/general/constructor_too_long.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+This might be acceptable, but it could be an indication that the constructor is doing too much work directly and should be delegating some of that work to other methods.
+A general rule is to try to keep everything inside a constructor or method at one level of abstraction.
+For example, input validation can be delegated to another method so that the constructor can focus on storing variables.
+
+Consider breaking down the constructor into calls to other semantically meaningful methods.
```
Comment (tehsphinx) [inline automated-comments/java/general/constructor_too_long.md]: potentially start with this. From the [guidelines](https://github.com/exercism/automated-analysis/blob/5b01bae1faced4c771308bab10a4fd6277e2a4d1/docs/comment_copy_guidelines.md):

> When possible, put the recommendation first, then the explanation.

File: automated-comments/java/general/method_too_long.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+This might be acceptable, but it could be an indication that the method is doing too much work directly and should be delegating some of that work to other methods.
+A general rule is to try to keep everything inside a constructor or method at one level of abstraction.
+For example, iterating through the elements of a loop could be in one method while manipulating each individual element could be in another.
+
+Consider breaking down the methods into calls to other semantically meaningful methods.
```
Comment (tehsphinx) [inline automated-comments/java/general/method_too_long.md]: potentially start with this. From the [guidelines](https://github.com/exercism/automated-analysis/blob/5b01bae1faced4c771308bab10a4fd6277e2a4d1/docs/comment_copy_guidelines.md):

> When possible, put the recommendation first, then the explanation.

File: automated-comments/java/hamming/avoid_character_literals.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+This solution used character literals (eg. `'T'`), but they are not necessary for this exercise.
+Although the description mentions the DNA nucleotide use-case, you can calculate the Hamming distance for any two strings of the same length.
+Check if the letters are different, regardless of what the letters are.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/avoid_character_literals.md]: potentially start with this. From the [guidelines](https://github.com/exercism/automated-analysis/blob/5b01bae1faced4c771308bab10a4fd6277e2a4d1/docs/comment_copy_guidelines.md):

> When possible, put the recommendation first, then the explanation.

File: automated-comments/java/hamming/calculate_distance_in_constructor.md [hunk 10]
```diff
@@ -0,0 +1,3 @@
+This solution seems to be calculating the Hamming distance inside the `getHammingDistance()` method.
+While this is acceptable, it means that the calculation might be done again if you did not use lazy initialization.
+Consider calculating the Hamming distance inside the constructor and storing the result in a member variable which can be returned inside `getHammingDistance()`.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/calculate_distance_in_constructor.md]: potentially start with this. From the [guidelines](https://github.com/exercism/automated-analysis/blob/5b01bae1faced4c771308bab10a4fd6277e2a4d1/docs/comment_copy_guidelines.md):

> When possible, put the recommendation first, then the explanation.

File: automated-comments/java/hamming/calculate_distance_in_constructor.md [hunk 11]
```diff
@@ -0,0 +1,3 @@
+This solution seems to be calculating the Hamming distance inside the `getHammingDistance()` method.
+While this is acceptable, it means that the calculation might be done again if you did not use lazy initialization.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/calculate_distance_in_constructor.md]: Possibly rephrase without `While this is acceptable,`

See guidelines about "This is perfectly fine" example.

File: automated-comments/java/hamming/must_calculate_hamming_distance.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+There are no loops or lambdas in this solution.
```
Comment (tehsphinx) [inline automated-comments/java/hamming/must_calculate_hamming_distance.md]:
```suggestion
```

Comment (tehsphinx) [reply automated-comments/java/hamming/must_calculate_hamming_distance.md]: Could be removed as this is more of an explanation what the analyzer is checking for.

File: automated-comments/java/general/constructor_too_long.md (hunk 1 above)
Comment (ErikSchierboom) [reply automated-comments/java/general/constructor_too_long.md]: And perhaps also explain what is meant by "long" (too many arguments, arguments with large names, too much lines of code)?

File: automated-comments/java/general/method_too_long.md (hunk 2 above)
Comment (ErikSchierboom) [reply automated-comments/java/general/method_too_long.md]: And perhaps also explain what is meant by "long" (too many arguments, arguments with large names, too much lines of code)?

File: automated-comments/java/hamming/avoid_character_literals.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+This solution used character literals (eg. `'T'`), but they are not necessary for this exercise.
```
Comment (ErikSchierboom) [inline automated-comments/java/hamming/avoid_character_literals.md]:
```suggestion
This solution uses character literals (eg. `'T'`), but they are not necessary for this exercise.
```

File: automated-comments/java/hamming/must_use_constructor.md [hunk 14]
```diff
@@ -0,0 +1,8 @@
+This exercise must use a constructor that takes two Strings:
+```java
+Hamming(String leftStrand, String rightStrand) {
+    ...
+}
+```
+
+NOTE: do not modify the tests to allow alternative inputs and make sure the tests are passing.
```
Comment (ErikSchierboom) [inline automated-comments/java/hamming/must_use_constructor.md]: Is this something we can detect? It is perhaps not very user-friendly to display this message if the user didn't actually modify the tests.

File: automated-comments/java/hamming/must_use_string_char_at_or_code_point_at.md [hunk 15]
```diff
@@ -0,0 +1,2 @@
+When comparing characters to calculate the Hamming distance, use the String `charAt` or `codePointAt` method.
```
Comment (ErikSchierboom) [inline automated-comments/java/hamming/must_use_string_char_at_or_code_point_at.md]:
```suggestion
When comparing characters to calculate the Hamming distance, use the String's `charAt` or `codePointAt` method.
```

File: automated-comments/java/hamming/should_use_stream_filter_and_count.md [hunk 16]
```diff
@@ -0,0 +1 @@
+Stream-based computations should make use of Stream's `filter` and `count` methods.
```
Comment (ErikSchierboom) [inline automated-comments/java/hamming/should_use_stream_filter_and_count.md]: I'm wondering about "should." As a student, I might be wondering if I was expected to change anything or not. Is the student required to make a change? If so, maybe rephrase to something like the comment below.

File: automated-comments/java/hamming/should_use_string_is_empty.md [hunk 17]
```diff
@@ -0,0 +1 @@
+Use the String's `isEmpty()` method to check if the string is empty.
```
Comment (ErikSchierboom) [inline automated-comments/java/hamming/should_use_string_is_empty.md]: Nice and clear comment! 👍

File: automated-comments/java/hamming/must_use_constructor.md (hunk 14 above)
Comment (SleeplessByte) [reply automated-comments/java/hamming/must_use_constructor.md]: Since the disapprovals are passed to a mentor, you don't really need to guard for that. I think you _may_ assume the tests are unaltered, for now.

**Comment by ErikSchierboom** [comment]
@jmrunkle I've re-requested a review from @SleeplessByte and @tehsphinx .
